### PR TITLE
Support type alias as Literal

### LINKED ISSUE #7996: Variant of alias for enum type cannot be used as Literal
```python
class Foo(Enum):
    A = 1

F = Foo

# works
x: Literal[Foo.A]
# name-defined error
y: Literal[F.A]
```

Is there a simple workaround here? I also tried `F: Final = Foo` which doesn't work either

### PR DESCRIPTION
**Pull request by TH3CHARLie**
resolves #7996

I'm not very sure about how effective the current testcase is, suggestions are welcomed.

### CODE REVIEW

File: test-data/unit/check-literal.test [hunk 1]
```diff
@@ -3149,3 +3149,18 @@ err_code = -TWO
...
+    class In:
+        class Inner:
+            pass
+
+O = Out
+I = Out.In
+OI = O.In
+x: O.In
+y: I.Inner
+z: OI.Inner
```
Comment (ilevkivskyi) [inline test-data/unit/check-literal.test]: Please add a `reveal_type()` for all three variables. (Also please keep newline at the end of file.)

File: test-data/unit/check-literal.test [hunk 2]
```diff
@@ -3149,3 +3149,18 @@ err_code = -TWO
 if bool():
     err_code = -THREE
 [builtins fixtures/float.pyi]
+
+[case testAliasForEnumTypeAsLiteral]
```
Comment (ilevkivskyi) [inline test-data/unit/check-literal.test]: Could you please actually add a test with enum and literal (as originally reported in the issue)?

File: mypy/semanal.py [hunk 3]
```diff
@@ -3969,6 +3969,12 @@ def lookup_qualified(self, name: str, ctx: Context,
                     namespace = node.fullname
                 elif isinstance(node, PlaceholderNode):
                     return sym
+                elif (isinstance(node, TypeAlias) and
+                      isinstance(get_proper_type(node.target), Instance)):
```
Comment (ilevkivskyi) [inline mypy/semanal.py]: I think this should be supported only for `no_args` aliases, i.e. you can almost literally copy the code from `visit_member_expr()`. Btw, it seems to me the TODO there is no more relevant, you can also try adding a test for situation like `A = Out; B = A; x: B.In`.

**Comment by TH3CHARLie** [comment]
Updated:
- `reveal_type()` for all declared vars
- rename original test to  `testAliasAsLiteral` (I forgot to change the name when I decided to exclude the enum test)
- add enum testcase `testAliasForEnumTypeAsLiteral`
- update `lookup_qualified` logic, remove TODO in `visit_member_expr`

File: test-data/unit/check-literal.test [hunk 4]
```diff
@@ -3149,3 +3149,40 @@ err_code = -TWO
 if bool():
     err_code = -THREE
 [builtins fixtures/float.pyi]
+
+[case testAliasAsLiteral]
```
Comment (ilevkivskyi) [inline test-data/unit/check-literal.test]:
```suggestion
[case testNestedClassOnAliasAsType]
```

File: mypy/semanal.py [hunk 5]
```diff
@@ -3969,6 +3968,11 @@ def lookup_qualified(self, name: str, ctx: Context,
                     namespace = node.fullname
                 elif isinstance(node, PlaceholderNode):
                     return sym
+                elif isinstance(node, TypeAlias) and node.no_args:
+                    assert isinstance(node.target, ProperType)
+                    if isinstance(node.target, Instance):
+                        node = node.target.type
+                        nextsym = node.get(part)
```
Comment (ilevkivskyi) [inline mypy/semanal.py]: I don't think you need to introduce the temporary `node` here.

File: test-data/unit/check-literal.test [hunk 6]
```diff
@@ -3149,3 +3149,40 @@ err_code = -TWO
 if bool():
     err_code = -THREE
 [builtins fixtures/float.pyi]
+
+[case testNestedClassOnAliasAsType]
+from typing_extensions import Literal
```
Comment (ilevkivskyi) [inline test-data/unit/check-literal.test]: Actually one more thing, sorry for not noticing this before. This import is unused, and the test case can be moved to `check-type-aliases.test`.

Comment (TH3CHARLie) [reply test-data/unit/check-literal.test]: Wow, I didn't notice that unused Literal import either, great catch!

**Pull request merged**